test/cpp/microbenchmarks/bm_threadpool.cc

```diff
@@ -0,0 +1,387 @@
+
+static void BM_ThreadPool16AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 16);
+}
+BENCHMARK(BM_ThreadPool16AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool32AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 32);
+}
+BENCHMARK(BM_ThreadPool32AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool64AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 64);
+}
+BENCHMARK(BM_ThreadPool64AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool128AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 128);
+}
+BENCHMARK(BM_ThreadPool128AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool512AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 512);
+}
+BENCHMARK(BM_ThreadPool512AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool2048AddSelf(benchmark::State& state) {
+  ThreadPoolAddSelfHelper(state, 2048);
+}
+BENCHMARK(BM_ThreadPool2048AddSelf)->RangePair(524288, 524288, 1, 1024);
+
+#if defined(__GNUC__) && !defined(SWIG)
+#if defined(__i386__) || defined(__x86_64__)
+#define CACHELINE_SIZE 64
+#elif defined(__powerpc64__)
+#define CACHELINE_SIZE 128
+#elif defined(__aarch64__)
+#define CACHELINE_SIZE 64
+#elif defined(__arm__)
+#if defined(__ARM_ARCH_5T__)
+#define CACHELINE_SIZE 32
+#elif defined(__ARM_ARCH_7A__)
+#define CACHELINE_SIZE 64
+#endif
+#endif
+#ifndef CACHELINE_SIZE
+#define CACHELINE_SIZE 64
+#endif
+#endif
+
+// A functor (closure) that simulates closures with small but non-trivial amount
+// of work.
+class ShortWorkFunctorForAdd
+    : public grpc_experimental_completion_queue_functor {
+ public:
+  BlockingCounter* counter_;
+
+  ShortWorkFunctorForAdd() {
+    functor_run = &ShortWorkFunctorForAdd::Run;
+    internal_next = this;
+    internal_success = 0;
+    val_ = 0;
+  }
+  static void Run(grpc_experimental_completion_queue_functor* cb, int ok) {
+    auto* callback = static_cast<ShortWorkFunctorForAdd*>(cb);
+    // Uses pad to avoid compiler complaining unused variable error.
+    callback->pad[0] = 0;
+    for (int i = 0; i < 1000; ++i) {
+      callback->val_++;
+    }
+    callback->counter_->DecrementCount();
+  }
+
+ private:
+  char pad[CACHELINE_SIZE];
+  volatile int val_;
+};
+
+// Simulates workloads where many short running callbacks are added to the
+// threadpool. The callbacks are not enough to keep all the workers busy
+// continuously so the number of workers running changes overtime.
+//
+// In effect this tests how well the threadpool avoids spurious wakeups.
+static void BM_SpikyLoad(benchmark::State& state) {
+  const int num_threads = state.range(0);
+
+  const int kNumSpikes = 1000;
+  const int batch_size = 3 * num_threads;
+  std::vector<ShortWorkFunctorForAdd> work_vector(batch_size);
+  while (state.KeepRunningBatch(kNumSpikes * batch_size)) {
+    grpc_core::ThreadPool pool(num_threads);
```

Comment:
This test is creating a new ThreadPool in a loop? I'd expect that to be potentially expensive enough that it would add noise we care about.